Finished replicating all the Enumerable methods from The Odin Project

voscarmv opened the pull request:
For this project, I:
- Used regular expressions to completely and accurately replicate the #each method.
- Used procs and blocks as required by The Odin Project specifications.
- Tested the accuracy of my replicas by comparing their output to output from the original methods.

voscarmv committed:
Repurposed my_select again to create my_count.

voscarmv committed:
Added proc functionality to my_map.

voscarmv committed:
Finished inject and multiply_els methods.

voscarmv committed:
Added test cases with original Enumerable methods to compare results.

voscarmv committed:
Linted with rubocop.

voscarmv committed:
Updated README.md

voscarmv commented:
NOTE: The branch shows conflicts because I made a mistake at the beginning of the project: I made some commits by accident. I tried to correct this by swapping the master branch and continue working on the feature_branch.

When I created this pull request, the "branch conflicts" appeared. I have not been able to solve this problem.

However, if you approve this project, I will be able to merge the branch with the develop branch by resolving these conflicts. I can not resolve the conflicts before merging, so I will need your authorization to be able to merge, in order to resolve the conflicts afterward.

Please tell me if this is OK, or if I will have to create a new repository altogether.

voscarmv commented:
I can resolve the conflicts on this page, but it would require me to merge the changes, and I can only do that after the project is approved. Please tell me if I can do this, or if it would be better to just create a new repository altogether.
![image](https://user-images.githubusercontent.com/2739245/75514943-5b618f00-59be-11ea-9fed-961bfc1a884d.png)

pelzolga123 reviewed (changes requested):
Project requires changes.

pelzolga123 commented:
## Hi Oscar! ![](https://user-images.githubusercontent.com/38801423/62908122-703d4480-bd44-11e9-9115-42025ab7f341.png) 
### Your did a good job so far!
Here are the methods, that require fixing:
* [x] `my_all`:
* When no block or argument is given returns true when none of the collection members are false or nil:
` [1,2,3,4,5].my_all? ` - should return true
* When a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern:
` [1, 1, 1].my_all?(1) ` - should return true
* [x] `my_any?`:
Is not present in your project.
* [x] `my_none?`:
*  When no block or argument is given returns true only if none of the collection members is true
`[1, 2 ,3].my_none?` - should return false
* When a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
`[1,2,3,4,5].my_none?(2) - should return false
* [x] `my_inject`:
* Combines all elements of enum by applying a binary operation, specified by a block
` (5..10).my_inject { |sum, n| sum + n } ` - should return 45
* When a symbol is specified combines each element of the collection by applying the symbol as a named method
`puts (5..10).my_inject(:*)` - should return 151200
* [x] Comments: It’s better to delete commented code. If you need to review a previous version of your code, use git history ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/no_inline_comments.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))

### Please kindly fix the above issues and send another code review request.

If you have any questions or need help, you can message me on slack (@ Olga Pelts). *Happy coding*

voscarmv committed:
Corrections applied.

whiz25 reviewed (changes requested):
## Nice Progress :+1: 
You have made a tremendous progress in this project. Your code only has one failing edge case. Fix that so we can merge.
Check my comments for suggestions.

### Martin Afani

whiz25 commented on main.rb:
```diff
@@ -1,113 +1,222 @@
-      self.to_enum
+      arr.to_enum
+    end
+  end
```
- [x] should be identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] my_each_with_index calls the given block once for each element in self
```ruby
      array = Array.new(100) { rand(0...9) } }
      my_each_output = ''
      block = proc { |num, idx| my_each_output += "Num: #{num}, idx: #{idx}\n" }
      array.each_with_index(&block)
      each_output = my_each_output.dup
      my_each_output = ''
      array.my_each_with_index(&block)
      my_each_output == each_output #true
```

voscarmv committed:
Corrected edge case and added test to end of code.

voscarmv commented:
Thanks @whiz25. I've corrected the code and added your test case. It now returns true. Note: I had to add one space character to the `my_each_output` variable, because rubocop wouldn't let me add empty strings in variables.

```
main.rb(main):225:0> array = Array.new(100) { rand(0...9) }
=> [7, 5, 6, 4, 7, 4, 1, 2, 3, 3, 8, 1, 0, 6, 0, 4, 4, 5, 4, 6, 5, 6, 4, 6, 6, 0, 1, 4, 0, 8, 0, 0, 0, 8, 3, 1, 2, 3, 0, 5, 6, 7, 2, 6, 7, 6, 6, 0, 4, 1, 7, 4, 0, 2, 3, 1, 7, 1, 3, 4, 5, 7, 0, 7, 7, 3, 1, 2, 4, 7, 5, 1, 8, 2, 2, 6, 3, 8, 8, 1, 1, 1, 6, 5, 0, 0, 7, 3, 4, 7, 3, 6, 0, 0, 3, 5, 0, 7, 2, 6]
main.rb(main):226:0> my_each_output = ' '
=> " "
main.rb(main):227:0> block = proc { |num, idx| my_each_output += "Num: #{num}, idx: #{idx}\n" }
=> #<Proc:0x0000557a1645c8c0@main.rb:227>
main.rb(main):228:0> array.each_with_index(&block)
=> [7, 5, 6, 4, 7, 4, 1, 2, 3, 3, 8, 1, 0, 6, 0, 4, 4, 5, 4, 6, 5, 6, 4, 6, 6, 0, 1, 4, 0, 8, 0, 0, 0, 8, 3, 1, 2, 3, 0, 5, 6, 7, 2, 6, 7, 6, 6, 0, 4, 1, 7, 4, 0, 2, 3, 1, 7, 1, 3, 4, 5, 7, 0, 7, 7, 3, 1, 2, 4, 7, 5, 1, 8, 2, 2, 6, 3, 8, 8, 1, 1, 1, 6, 5, 0, 0, 7, 3, 4, 7, 3, 6, 0, 0, 3, 5, 0, 7, 2, 6]
main.rb(main):229:0> each_output = my_each_output.dup
=> " Num: 7, idx: 0\nNum: 5, idx: 1\nNum: 6, idx: 2\nNum: 4, idx: 3\nNum: 7, idx: 4\nNum: 4, idx: 5\nNum: 1, idx: 6\nNum: 2, idx: 7\nNum: 3, idx: 8\nNum: 3, idx: 9\nNum: 8, idx: 10\nNum: 1, idx: 11\nNum: 0, idx: 12\nNum: 6, idx: 13\nNum: 0, idx: 14\nNum: 4, idx: 15\nNum: 4, idx: 16\nNum: 5, idx: 17\nNum: 4, idx: 18\nNum: 6, idx: 19\nNum: 5, idx: 20\nNum: 6, idx: 21\nNum: 4, idx: 22\nNum: 6, idx: 23\nNum: 6, idx: 24\nNum: 0, idx: 25\nNum: 1, idx: 26\nNum: 4, idx: 27\nNum: 0, idx: 28\nNum: 8, idx: 29\nNum: 0, idx: 30\nNum: 0, idx: 31\nNum: 0, idx: 32\nNum: 8, idx: 33\nNum: 3, idx: 34\nNum: 1, idx: 35\nNum: 2, idx: 36\nNum: 3, idx: 37\nNum: 0, idx: 38\nNum: 5, idx: 39\nNum: 6, idx: 40\nNum: 7, idx: 41\nNum: 2, idx: 42\nNum: 6, idx: 43\nNum: 7, idx: 44\nNum: 6, idx: 45\nNum: 6, idx: 46\nNum: 0, idx: 47\nNum: 4, idx: 48\nNum: 1, idx: 49\nNum: 7, idx: 50\nNum: 4, idx: 51\nNum: 0, idx: 52\nNum: 2, idx: 53\nNum: 3, idx: 54\nNum: 1, idx: 55\nNum: 7, idx: 56\nNum: 1, idx: 57\nNum: 3, idx: 58\nNum: 4, idx: 59\nNum: 5, idx: 60\nNum: 7, idx: 61\nNum: 0, idx: 62\nNum: 7, idx: 63\nNum: 7, idx: 64\nNum: 3, idx: 65\nNum: 1, idx: 66\nNum: 2, idx: 67\nNum: 4, idx: 68\nNum: 7, idx: 69\nNum: 5, idx: 70\nNum: 1, idx: 71\nNum: 8, idx: 72\nNum: 2, idx: 73\nNum: 2, idx: 74\nNum: 6, idx: 75\nNum: 3, idx: 76\nNum: 8, idx: 77\nNum: 8, idx: 78\nNum: 1, idx: 79\nNum: 1, idx: 80\nNum: 1, idx: 81\nNum: 6, idx: 82\nNum: 5, idx: 83\nNum: 0, idx: 84\nNum: 0, idx: 85\nNum: 7, idx: 86\nNum: 3, idx: 87\nNum: 4, idx: 88\nNum: 7, idx: 89\nNum: 3, idx: 90\nNum: 6, idx: 91\nNum: 0, idx: 92\nNum: 0, idx: 93\nNum: 3, idx: 94\nNum: 5, idx: 95\nNum: 0, idx: 96\nNum: 7, idx: 97\nNum: 2, idx: 98\nNum: 6, idx: 99\n"
main.rb(main):230:0> my_each_output = ' '
=> " "
main.rb(main):231:0> array.my_each_with_index(&block)
=> nil
main.rb(main):232:0> my_each_output == each_output # true
=> true
main.rb(main):233:0> 
```

whiz25 reviewed (approved):
## Congz :clap: 
You have made the suggested changes and it's therefore :+1: from me. Merge and proceed.

### Martin Afani

voscarmv commented:
That's great news @whiz25 , thanks for the review!

voscarmv committed:
Merge branch 'develop' into feature_branch
